# Research Challenge

<div style="background-color: #f8d7da; border-left: 6px solid #ccc; margin: 20px; padding: 15px;">
    <strong>💡 Margaret Atwood:</strong> Every aspect of human technology has a dark side, including the bow and arrow.
</div>

## 🏅 Build your own model

It is time to go back to supervised machine learning problems.

You have been assigned one dataset from [MatBench](https://matbench.materialsproject.org) as introduced in the [Lecture slides](https://speakerdeck.com/aronwalsh/mlformaterials-challenge-25). You are free to choose and tune any machine-learning model, with any Python library, but it should be appropriate for the problem. For instance, [XGBoost](https://xgboost.readthedocs.io) could be a good starting starting point to build a regression model. You can refer back to earlier notebooks and repurpose code as needed.

You may reach the limits of computing processing power on Google Colab. Building a useful model with limited resources is a real-world skill. Using other free resources is allowed if you find an alternative service, as is running on your own computer. A model tracker such as [wandb](https://wandb.ai) could be helpful for advanced users. If you want to try a brute force approach, a library such as [Automatminer](https://hackingmaterials.lbl.gov/automatminer) may be of interest.

This notebook should be used for keeping a record of your model development, submission, and even your presentation. You are free to edit (add/remove/delete) or rearrange the cells as you see fit.

### Your details

In [1]:
import numpy as np

# Insert your values
Name = "Mark Zangwill" # Replace with your name
CID = 2198282 # Replace with your College ID (as a numeric value with no leading 0s)

# Set a random seed using the CID value
CID = int(CID)
np.random.seed(CID)

# Print the message
print("This is the work of " + Name + " [CID: " + str(CID) + "]\n")

# Define the available groups
groups = ['A', 'B', 'C', 'D', 'E']

# Select a group based on the seeded random state
challenge_group = np.random.choice(groups)

# Print the challenge code
print("Your challenge code is " + challenge_group)

This is the work of Mark Zangwill [CID: 2198282]

Your challenge code is C


## Problem statement
You have been assigned one dataset from the [list](https://matbench.materialsproject.org/Benchmark%20Info/matbench_v0.1/) on [MatBench](https://matbench.materialsproject.org). You should state what problem you are trying to solve and comment on the best-performing model in the benchmark.

In [2]:
# Spare cell




## Data preparation

Check the data distribution and apply appropriate pre-processing steps as required.

In [3]:
# Installation of libraries
!pip install matminer # Datasets and featurisation

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 62.1/62.1 kB 1.3 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 55.6/55.6 kB 1.7 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 5.3/5.3 MB 22.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 52.7/52.7 kB 1.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 4.7/4.7 MB 37.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.7/1.7 MB 23.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 35.2/35.2 MB 15.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 331.1/331.1 kB 17.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 332.3/332.3 kB 11.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 118.1/118.1 kB 4.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 962.5/962.5 kB 30.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 60.1/60.1 kB 2.2 MB/s eta 0:00:00
  Created 

In [4]:
# Get dataset info from matminer
from matminer.datasets import get_all_dataset_info
from matminer.datasets import load_dataset

# Detailed on https://hackingmaterials.lbl.gov/matminer/dataset_summary.html
# Uncomment the info line for your assigned challenge code

  # A
#info = get_all_dataset_info("matbench_dielectric")

  # B
#info = get_all_dataset_info("matbench_expt_gap")

  # C
info = get_all_dataset_info("matbench_expt_is_metal")

  # D
#info = get_all_dataset_info("matbench_glass")

  # E
#info = get_all_dataset_info("matbench_steels")

# Check the dataset information
print(info)

Dataset: matbench_expt_is_metal
Description: Matbench v0.1 test dataset for classifying metallicity from composition alone. Retrieved from Zhuo et al. supplementary information. Deduplicated according to composition, ensuring no conflicting reports were entered for any compositions (i.e., no reported compositions were both metal and nonmetal). For benchmarking w/ nested cross validation, the order of the dataset must be identical to the retrieved data; refer to the Automatminer/Matbench publication for more details.
Columns:
	composition: Chemical formula.
	is_metal: Target variable. 1 if is a metal, 0 if nonmetal.
Num Entries: 4921
Reference: Y. Zhuo, A. Masouri Tehrani, J. Brgoch (2018) Predicting the Band Gaps of Inorganic Solids by Machine Learning J. Phys. Chem. Lett. 2018, 9, 7, 1668-1673 
 https//:doi.org/10.1021/acs.jpclett.8b00124.
Bibtex citations: ["@Article{Dunn2020,\nauthor={Dunn, Alexander\nand Wang, Qi\nand Ganose, Alex\nand Dopp, Daniel\nand Jain, Anubhav},\ntitle={Benc

In [5]:
# Load your dataset into a pandas DataFrame
df = load_dataset("matbench_expt_is_metal")

print(df)

Fetching matbench_expt_is_metal.json.gz from https://ml.materialsproject.org/projects/matbench_expt_is_metal.json.gz to /usr/local/lib/python3.12/dist-packages/matminer/datasets/matbench_expt_is_metal.json.gz


Fetching https://ml.materialsproject.org/projects/matbench_expt_is_metal.json.gz in MB: 0.034816MB [00:00, 42.46MB/s]                 

            composition  is_metal
0              Ag(AuS)2      True
1            Ag(W3Br7)2      True
2      Ag0.5Ge1Pb1.75S4     False
3     Ag0.5Ge1Pb1.75Se4     False
4                Ag2BBr      True
...                 ...       ...
4916             ZrTaN3     False
4917               ZrTe      True
4918             ZrTi2O      True
4919             ZrTiF6      True
4920               ZrW2      True

[4921 rows x 2 columns]


In [6]:
df.describe()

,composition,is_metal
count,4921,4921
unique,4921,2
top,ZrW2,False
freq,1,2470


Where middle column is the "composition" column, with all 4921 being unique. The last column beign the output, unique being two since outputs are either true or false, depending on whenther metal or non.

Choose relevant features, which may be based on composition or structure, depending on your problem. [matminer](https://hackingmaterials.lbl.gov/matminer/) is a good place to start.

###Materials Featurization

In [7]:
#download ElementEmbedings
!pip install elementembeddings

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.0/3.0 MB 13.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.5/3.5 MB 31.4 MB/s eta 0:00:00


Use magpie embeding schemes

In [21]:
# Featurise all chemical compositions
from elementembeddings.composition import composition_featuriser

def featurise(dataframe, embedding):

  # Compute element embeddings using mean and max pooling
  mean_df = composition_featuriser(dataframe["composition"], embedding=embedding, stats=["mean"])
  max_df = composition_featuriser(dataframe["composition"], embedding=embedding, stats=["maxpool"])
  total_df = composition_featuriser(dataframe["composition"], embedding=embedding, stats=["mean","variance", "maxpool","minpool", "maxpool", "sum"])

  # Convert "is_metal" column to integer labels (0, 1)
  dataframe['is_metal'] = dataframe['is_metal'].astype(int)
  mean_df['is_metal'] = dataframe['is_metal']
  max_df['is_metal'] = dataframe['is_metal']
  total_df['is_metal'] = dataframe['is_metal']

  # Define feature matrices and target variable
  cols_to_drop = ['is_metal', 'formula']

  X_mean = mean_df.drop(columns=cols_to_drop, errors='ignore').values
  X_max = max_df.drop(columns=cols_to_drop, errors='ignore').values
  X_total = total_df.drop(columns=cols_to_drop, errors='ignore').values
  y = dataframe['is_metal'].values  # Target variable

  return X_mean, X_max, mean_df, max_df, X_total, total_df, y

X_mean, X_max, mean_df, max_df, X_total, total_df, y = featurise(df, "magpie")

print("Mean pooling features (first two rows, first 4 columns):")
print(mean_df.iloc[:2, :4])
print("\nMax pooling features (first two rows, first 4 columns):")
print(max_df.iloc[:2, :4])
print(total_df.iloc[:2, :4])

Featurising compositions...


100%|██████████| 4921/4921 [00:14<00:00, 343.47it/s]


Computing feature vectors...


100%|██████████| 4921/4921 [00:00<00:00, 41725.97it/s]


Featurising compositions...


100%|██████████| 4921/4921 [00:13<00:00, 371.60it/s]


Computing feature vectors...


100%|██████████| 4921/4921 [00:00<00:00, 124852.07it/s]


Featurising compositions...


100%|██████████| 4921/4921 [00:12<00:00, 382.27it/s]


Computing feature vectors...


100%|██████████| 4921/4921 [00:00<00:00, 32620.66it/s]


Mean pooling features (first two rows, first 4 columns):
      formula  mean_Number  mean_MendeleevNumber  mean_AtomicWeight
0    Ag(AuS)2    47.400000                  74.6         113.186268
1  Ag(W3Br7)2    46.714286                  81.0         110.931629

Max pooling features (first two rows, first 4 columns):
      formula  maxpool_Number  maxpool_MendeleevNumber  maxpool_AtomicWeight
0    Ag(AuS)2            79.0                     88.0            196.966569
1  Ag(W3Br7)2            74.0                     95.0            183.840000
      formula  mean_Number  mean_MendeleevNumber  mean_AtomicWeight
0    Ag(AuS)2    47.400000                  74.6         113.186268
1  Ag(W3Br7)2    46.714286                  81.0         110.931629


In [16]:
total_df = composition_featuriser(df["composition"], embedding="magpie", stats=["mean","variance", "maxpool","minpool", "maxpool", "sum"])
print(total_df.shape[1])

Featurising compositions...


100%|██████████| 4921/4921 [00:13<00:00, 378.27it/s]


Computing feature vectors...


100%|██████████| 4921/4921 [00:00<00:00, 30957.64it/s]


133


In [9]:
#print the number columns in the dataframes
print(mean_df.shape[1])
print(max_df.shape[1])
#print the number of rows
print(mean_df.shape[0])
print(max_df.shape[0])

24
24
4921
4921


pick the best stats to use. e.g. mean or variance or ...

embedding scheme could also be tested...

In [ ]:
#determine the most important features
from sklearn.ensemble import DecisionTreeClassifier, RandomForestClassifier, GradientBoostingClassifier
from sklearn.feature_selection import SelectFromModel

randomForestClassifier = RandomForestClassifier(random_state=42)


## Model selection, testing and training

Define your model and justify your choice based on the problem and available data. You can look back at earlier notebooks and investigate other examples online including in [scikit-learn](https://scikit-learn.org).

In [10]:
#apply nested cross validation
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import  cross_val_score
from sklearn.metrics import roc_auc_score

decisionTreeClassifier = DecisionTreeClassifier(random_state= 42)
scores = cross_val_score(decisionTreeClassifier, X_mean, y, cv=5, scoring='roc_auc')
print("ROC-AUC scores:", scores)
print("mean CV accuray", scores.mean())

ROC-AUC scores: [0.81911451 0.82939767 0.83119061 0.80070644 0.84967363]
mean CV accuray 0.8260165722827468


In [11]:
from sklearn.ensemble import RandomForestClassifier

randomForestClassifier = RandomForestClassifier(random_state=42)
scores = cross_val_score(randomForestClassifier, X_mean, y, cv=5, scoring='roc_auc')
print("ROC-AUC scores:", scores)
print("mean CV accuray", scores.mean())

ROC-AUC scores: [0.94944012 0.95300132 0.97276915 0.9191089  0.95267083]
mean CV accuray 0.9493980637680008


In [12]:
# gradient boosted method
from sklearn.ensemble import GradientBoostingClassifier

gradientBoostingClassifier = GradientBoostingClassifier(random_state=42)
scores = cross_val_score(gradientBoostingClassifier, X_mean, y, cv=5, scoring='roc_auc')
print("ROC-AUC scores:", scores)
print("mean CV accuray", scores.mean())

ROC-AUC scores: [0.94248291 0.94673015 0.96578947 0.90432744 0.93843055]
mean CV accuray 0.9395521056283196


To conduct hyperparameter optimization, nested crossed validation is requred as to not leak data, giving optimisitic results. This involves having two loops; an inner and outer. The inner is where featurization is done as well as hyperparameter optimization. the outer does test.

Train, validate and test your model. Make sure to do proper data splits and to consider the hyperparamaters of your model.

<details>
<summary>Note on the ROC-AUC classification metric</summary>
There is one metric we didn't cover but is used in Matbench. In binary classification models, the ROC-AUC (Receiver Operating Characteristic - Area Under the Curve) score can be used to evaluate performance. It quantifies the ability of the model to distinguish between positive and negative instances across different decision thresholds. A higher ROC-AUC score (ranging from 0.5 to 1) indicates better performance, with 1 representing a perfect classifier and 0.5 indicating performance no better than random chance. There is a more detailed discussion here: https://developers.google.com/machine-learning/crash-course/classification/roc-and-auc.

The metric can be calculated using the `roc_auc_score` function from the `sklearn.metrics` module, e.g.

```python
from sklearn.metrics import roc_auc_score

# Assuming you have true labels (y_true) and predicted probabilities (y_pred_prob)
y_true = [...]  
y_pred_prob = [...]  

# Calculate ROC-AUC
roc_auc = roc_auc_score(y_true, y_pred_prob)

# Display the result
print(f'ROC-AUC Score: {roc_auc:.4f}')
```
</details>

In [13]:
# Spare cell




## Model analysis and discussion

How well does your final model perform? Think of metrics and plots that are useful to dig a little deeper.

Compare against the best-performing model on the [MatBench](https://matbench.materialsproject.org) leaderboard.  With limited resources, don't expect to match this performance, but you should do better than a baseline model.

In [14]:
# Spare cell




## Large Language Model (LLM) usage declaration

Acknowledge use of a generative model during your assignment. Points to consider:

* State which LLM (e.g. GPT-4, Gemini, Co-Pilot)

* Specify tasks (e.g. summarising research or code snippets)

* Were any limitations/biases noted?

* How did you ensure ethical use?

In [15]:
# Spare cell


